haystack/generator/transformers.py

```diff
@@ -291,3 +292,181 @@ def predict(self, query: str, documents: List[Document], top_k: Optional[int] =
+        :param use_gpu: Whether to use GPU (if available)
+        """
+
+        self.model_name_or_path = model_name_or_path
+        self.max_length = max_length
+        self.min_length = min_length
+        self.num_beams = num_beams
+
+        if top_k > self.num_beams:
+            top_k = self.num_beams
+            logger.warning(f'top_k value should not be greater than num_beams, hence setting it to {num_beams}')
+
+        self.top_k = top_k
+
+        if use_gpu and torch.cuda.is_available():
+            self.device = torch.device("cuda")
+        else:
+            self.device = torch.device("cpu")
+
+        Seq2SeqGenerator._register_converters(model_name_or_path, input_converter)
+
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        self.model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path).to(self.device)
+        self.model.eval()
+
+    @classmethod
+    def _register_converters(cls, model_name_or_path: str, custom_converter: Optional[Callable]):
+        # init if empty
+        if not cls._model_input_converters:
+            cls._model_input_converters["yjernite/bart_eli5"] = _BartEli5Converter()
+
+        # register user provided custom converter
+        if model_name_or_path and custom_converter:
+            cls._model_input_converters[model_name_or_path] = custom_converter
+
+    @classmethod
+    def _get_converter(cls, model_name_or_path: str):
+        return cls._model_input_converters.get(model_name_or_path)
+
+    def predict(self, query: str, documents: List[Document], top_k: Optional[int] = None) -> Dict:
+        """
+        Generate the answer to the input query. The generation will be conditioned on the supplied documents.
+        These document can be retrieved via the Retriever or supplied directly via predict method.
+
+        :param query: Query
+        :param documents: Related documents (e.g. coming from a retriever) that the answer shall be conditioned on.
+        :param top_k: Number of returned answers
+        :return: Generated answers
+
+        """
+        torch.set_grad_enabled(False)
+        if len(documents) == 0:
+            raise AttributeError("generator needs documents to predict the answer")
+
+        top_k = top_k if top_k is not None else self.top_k
+
+        if top_k > self.num_beams:
+            top_k = self.num_beams
+            logger.warning(f'top_k value should not be greater than num_beams, hence setting it to {top_k}')
+
+        converter: Callable = Seq2SeqGenerator._get_converter(self.model_name_or_path)
+        if not converter:
+            raise KeyError(f"Seq2SeqGenerator doesn't have input converter registered for {self.model_name_or_path}. "
+                           f"Provide custom converter for {self.model_name_or_path} in Seq2SeqGenerator initialization")
+
+        query_and_docs_encoded = converter(self.tokenizer, query, documents, top_k).to(self.device)
```

Comment:
How to make sure user supplied Callable follow this syntax?
I understand Callable is very elegant but these is chance of error.
How about having Base class which have abstract function with this signature?

Reply:
When you say "chance of error" are you referring to a user potentially mismatching the signature of Callable? I can see that point but we'd like to keep this Callable relatively hidden and for super-user customization. Your call @tholor

Reply:
I would prefer going forward with the light-weight Callable. However, we should add the expected signature to the docstring of  the `Seq2SeqGenerator.__init__()` and maybe link `_BartEli5Converter` as an example.

To make it a bit more failsafe, we could change the call to `converter(tokenizer=self.tokenizer, query=query, documents=documents, top_k=top_k).to(self.device)` and catch the `TypeError` with a descriptive error message if the converter doesn't support this signature :thinking:

Reply:
Great suggestion @tholor thank you